apps/wallet/src/background/keyring/VaultStorage.ts

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import { randomBytes } from '@noble/hashes/utils';
+import Browser from 'webextension-polyfill';
+
+import { Vault } from './Vault';
+import { getRandomEntropy, toEntropy } from '_shared/utils/bip39';
+
+import type { StoredData } from './Vault';
+import type { Storage } from 'webextension-polyfill';
+
+export const IS_SESSION_STORAGE_SUPPORTED = 'chrome' in global;
+const SESSION_STORAGE: Storage.LocalStorageArea | null =
+    // @ts-expect-error chrome
+    IS_SESSION_STORAGE_SUPPORTED ? global.chrome.storage.session : null;
+const LOCAL_STORAGE = Browser.storage.local;
+
+// we use this password + a random one for each time we store the encrypted
+// vault to session storage
+const PASSWORD =
+    process.env.WALLET_KEYRING_PASSWORD ||
+    '344c6f7d04a65c24f35f5c710b0e91e2f2e2f88c038562622d5602019b937bc2c2aa2821e65cc94775fe5acf2fee240d38f1abbbe00b0e6682646a4ce10e908e';
+const VAULT_KEY = 'vault';
+const EPHEMERAL_PASSWORD_KEY = '244e4b24e667ebf';
+const EPHEMERAL_VAULT_KEY = 'a8e451b8ae8a1b4';
+
+async function getFromStorage<T>(
+    storage: Storage.LocalStorageArea,
+    key: string
+): Promise<T | null> {
+    return (await storage.get({ [key]: null }))[key];
+}
+
+async function setToStorage<T>(
+    storage: Storage.LocalStorageArea,
+    key: string,
+    value: T
+): Promise<void> {
+    return await storage.set({ [key]: value });
+}
+
+async function ifSessionStorage(
```

Comment:
I don't love this API of using a callback, I feel like this would be cleaner as an inline `if` statement.

Reply:
I had it like that and thought this was cleaner! 😄 I will update it

Reply:
Ah hah I don't feel super strongly, so you can keep it if you want.

Reply:
merging it as it is then 👍